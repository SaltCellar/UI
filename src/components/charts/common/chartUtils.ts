import { Report } from 'api/reports/report';
import endOfMonth from 'date-fns/end_of_month';
import format from 'date-fns/format';
import formatDate from 'date-fns/format';
import getDate from 'date-fns/get_date';
import getYear from 'date-fns/get_year';
import startOfMonth from 'date-fns/start_of_month';
import {
  ComputedReportItem,
  getComputedReportItems,
} from 'utils/computedReport/getComputedReportItems';
import {
  FormatOptions,
  unitLookupKey,
  ValueFormatter,
} from 'utils/formatValue';
import { SortDirection } from 'utils/sort';
import { t } from '../../i18nProvider';

export interface ChartDatum {
  childName?: string;
  key: string | number;
  name?: string | number;
  show?: boolean;
  tooltip?: string;
  units: string;
  x: string | number;
  y: number;
}

// The computed report cost or usage item
export const enum ComputedReportItemType {
  cost = 'cost', // cost.total.value
  infrastructure = 'infrastructure', // infrastructure.total.value
  supplementary = 'supplementary', // supplementary.total.value
  usage = 'usage', // usage.value
}

// The computed report value
export const enum ComputedReportItemValueType {
  none = 'none', // A value type is not used in this scenario (e.g., usage.value)
  markup = 'markup', // infrastructure.markup.value
  raw = 'raw', // infrastructure.raw.value
  total = 'total', // // infrastructure.total.value
  usage = 'usage', // infrastructure.usage.value
}

export const enum ChartType {
  rolling,
  daily,
  monthly,
}

export function transformReport(
  report: Report,
  type: ChartType = ChartType.daily,
  key: any = 'date',
  reportItem: string = 'cost',
  reportItemValue: string = 'total'
): ChartDatum[] {
  if (!report) {
    return [];
  }
  const items = {
    idKey: key,
    report,
    reportItemValue,
    sortKey: 'id',
    sortDirection: SortDirection.desc,
  } as any;
  const computedItems = getComputedReportItems(items);
  if (type === ChartType.daily) {
    return computedItems.map(i =>
      createDatum(i[reportItem], i, key, reportItem)
    );
  }
  if (type === ChartType.monthly) {
    return computedItems.map(i =>
      createDatum(i[reportItem], i, key, reportItem)
    );
  }
  return computedItems.reduce<ChartDatum[]>((acc, d) => {
    const prevValue = acc.length ? acc[acc.length - 1].y : 0;
    return [...acc, createDatum(prevValue + d[reportItem], d, key, reportItem)];
  }, []);
}

export function createDatum<T extends ComputedReportItem>(
  value: number,
  computedItem: T,
  idKey = 'date',
  reportItem: string = 'cost'
): ChartDatum {
  const xVal = idKey === 'date' ? getDate(computedItem.id) : computedItem.label;
  const yVal = isFloat(value)
    ? parseFloat(value.toFixed(2))
    : isInt(value)
    ? value
    : 0;
  return {
    x: xVal,
    y: yVal,
    key: computedItem.id,
    name: computedItem.id,
    units: computedItem.units[reportItem],
  };
}

export function getDatumDateRange(
  datums: ChartDatum[],
  offset: number = 0
): [Date, Date] {
  if (!(datums && datums.length)) {
    const today = new Date();

    // If datums is empty, obtain the month based on offset (e.g., to show previous month in chart legends)
    if (offset) {
      today.setMonth(today.getMonth() - offset);
    }
    const firstOfMonth = startOfMonth(today);
    return [firstOfMonth, today];
  }

  const start = new Date(datums[0].key + 'T00:00:00');
  const end = new Date(datums[datums.length - 1].key + 'T00:00:00');
  return [start, end];
}

export function getDateRange(
  datums: ChartDatum[],
  firstOfMonth: boolean = true,
  lastOfMonth: boolean = false,
  offset: number = 0
): [Date, Date] {
  const [start, end] = getDatumDateRange(datums, offset);

  // Show the date range we are trying to cover (i.e., days 1-30/31)
  if (firstOfMonth && start.setDate) {
    start.setDate(1);
  }
  if (lastOfMonth && start.setDate) {
    const lastDate = endOfMonth(start).getDate();
    end.setDate(lastDate);
  }
  return [start, end];
}

export function getDateRangeString(
  datums: ChartDatum[],
  firstOfMonth: boolean = false,
  lastOfMonth: boolean = false,
  offset: number = 0
) {
  const [start, end] = getDateRange(datums, firstOfMonth, lastOfMonth, offset);

  return t(`chart.date_range`, {
    count: getDate(end),
    endDate: formatDate(end, 'DD'),
    month: Number(formatDate(start, 'M')) - 1,
    startDate: formatDate(start, 'DD'),
    year: getYear(end),
  });
}

export function getMonthRangeString(
  datums: ChartDatum[],
  key: string = 'chart.month_legend_label',
  offset: number = 0
): [string, string] {
  const [start, end] = getDateRange(datums, true, false, offset);

  return [
    t(key, {
      month: Number(formatDate(start, 'M')) - 1,
    }),
    t(key, {
      month: Number(formatDate(end, 'M')) - 1,
    }),
  ];
}

export function getMaxValue(datums: ChartDatum[]) {
  let max = 0;
  if (datums && datums.length) {
    datums.forEach(datum => {
      if (datum.y > max) {
        max = datum.y;
      }
    });
  }
  return max;
}

export function getTooltipContent(formatValue) {
  return function labelFormatter(
    value: number,
    unit: string = null,
    options: FormatOptions = {}
  ) {
    const lookup = unitLookupKey(unit);
    switch (lookup) {
      case 'core-hours':
      case 'hrs':
      case 'gb':
      case 'gb-hours':
      case 'gb-mo':
      case 'vm-hours':
        return t(`unit_tooltips.${lookup}`, {
          value: `${formatValue(value, unit, options)}`,
        });
      default:
        return `${formatValue(value, unit, options)}`;
    }
  };
}

export function getTooltipLabel(
  datum: ChartDatum,
  formatValue: ValueFormatter,
  formatOptions?: FormatOptions,
  idKey: any = 'date',
  units?: string
) {
  if (!datum.key) {
    return '';
  }
  if (idKey === 'date') {
    const date = format(datum.key, 'DD MMM YYYY');
    return `${date} ${formatValue(
      datum.y,
      units ? units : datum.units,
      formatOptions
    )}`;
  }
  return datum.key.toString();
}

export function getCostRangeString(
  datums: ChartDatum[],
  key: string = 'chart.cost_legend_label',
  firstOfMonth: boolean = false,
  lastOfMonth: boolean = false,
  offset: number = 0
) {
  const [start, end] = getDateRange(datums, firstOfMonth, lastOfMonth, offset);

  return t(key, {
    count: getDate(end),
    endDate: formatDate(end, 'D'),
    month: Number(formatDate(start, 'M')) - 1,
    startDate: formatDate(start, 'D'),
    year: getYear(end),
  });
}

export function getUsageRangeString(
  datums: ChartDatum[],
  key: string = 'chart.usage_legend_label',
  firstOfMonth: boolean = false,
  lastOfMonth: boolean = false,
  offset: number = 0
) {
  return getCostRangeString(datums, key, firstOfMonth, lastOfMonth, offset);
}

function isInt(n) {
  return Number(n) === n && n % 1 === 0;
}

function isFloat(n) {
  return Number(n) === n && n % 1 !== 0;
}
