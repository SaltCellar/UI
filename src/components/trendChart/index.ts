export { TrendChart, TrendChartProps } from './trendChart';
export { TrendChartDatum, DatumValueFormatter } from './trendChartUtils';
