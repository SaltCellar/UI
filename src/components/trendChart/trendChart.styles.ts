import { StyleSheet } from '@patternfly/react-styles';
import {
  global_primary_color_100,
  global_primary_color_200,
  global_success_color_100,
  global_success_color_200,
} from '@patternfly/react-tokens';
import { VictoryStyleInterface } from 'victory';

export const chartStyles = {
  colorScale: [global_primary_color_100.value, global_success_color_100.value],
  previousMonth: {
    data: {
      fill: global_success_color_200.value,
      stroke: global_success_color_100.value,
    },
  } as VictoryStyleInterface,
  currentMonth: {
    data: {
      fill: global_primary_color_100.value,
      stroke: global_primary_color_200.value,
    },
  } as VictoryStyleInterface,
};

export const styles = StyleSheet.create({
  reportSummaryTrend: {
    ':not(foo) svg': {
      overflow: 'visible',
    },
  },
});
