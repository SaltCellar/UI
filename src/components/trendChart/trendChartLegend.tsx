import { css } from '@patternfly/react-styles';
import { StyleSheet } from '@patternfly/react-styles';
import React from 'react';

const styles = StyleSheet.create({
  trendLegend: {
    display: 'flex',
    alignItems: 'center',
  },
});

interface TrendChartLegendProps {
  children: React.ReactNode;
}

const TrendChartLegend: React.SFC<TrendChartLegendProps> = ({ children }) => (
  <div className={css(styles.trendLegend)}>{children}</div>
);

export { TrendChartLegend, TrendChartLegendProps };
