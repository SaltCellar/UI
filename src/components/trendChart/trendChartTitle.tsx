import { css } from '@patternfly/react-styles';
import { StyleSheet } from '@patternfly/react-styles';
import { global_FontSize_xs, global_spacer_sm } from '@patternfly/react-tokens';
import React from 'react';

const styles = StyleSheet.create({
  trendChartTitle: {
    fontSize: global_FontSize_xs.value,
    marginBottom: global_spacer_sm.value,
  },
});

interface TrendChartTitleProps {
  children: React.ReactNode;
}

const TrendChartTitle: React.SFC<TrendChartTitleProps> = ({ children }) => (
  <div className={css(styles.trendChartTitle)}>{children}</div>
);

export { TrendChartTitle, TrendChartTitleProps };
