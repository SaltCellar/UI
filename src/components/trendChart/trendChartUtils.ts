import { Report } from 'api/reports';
import format from 'date-fns/format';
import getDate from 'date-fns/get_date';
import startOfMonth from 'date-fns/start_of_month';
import { FormatOptions, ValueFormatter } from 'utils/formatValue';
import {
  ComputedReportItem,
  getComputedReportItems,
} from 'utils/getComputedReportItems';
import { SortDirection } from 'utils/sort';

export interface TrendChartDatum {
  x: number;
  y: number;
  date: string | number;
  units: string;
}

export const enum TrendChartType {
  rolling,
  daily,
}

export function transformReport(
  report: Report,
  type: TrendChartType = TrendChartType.daily
): TrendChartDatum[] {
  if (!report) {
    return [];
  }

  const computedItems = getComputedReportItems({
    report,
    idKey: 'date',
    sortKey: 'id',
    sortDirection: SortDirection.desc,
  });

  if (type === TrendChartType.daily) {
    return computedItems.map(i => createDatum(i.total, i));
  }

  return computedItems.reduce<TrendChartDatum[]>((acc, d) => {
    const prevValue = acc.length ? acc[acc.length - 1].y : 0;
    return [...acc, createDatum(prevValue + d.total, d)];
  }, []);
}

export function createDatum(
  value: number,
  computedItem: ComputedReportItem
): TrendChartDatum {
  return {
    x: getDate(computedItem.id),
    y: value,
    date: computedItem.id,
    units: computedItem.units,
  };
}

export function getDatumDateRange(datums: TrendChartDatum[]): [Date, Date] {
  if (!datums.length) {
    const today = new Date();
    const firstOfMonth = startOfMonth(today);
    return [firstOfMonth, today];
  }

  const start = new Date(datums[0].date + 'T00:00:00');
  const end = new Date(datums[datums.length - 1].date + 'T00:00:00');
  return [start, end];
}

export function getDateRangeString(datums: TrendChartDatum[]) {
  const [start, end] = getDatumDateRange(datums);
  const monthName = format(start, 'MMM');
  const startDate = getDate(start);
  const endDate = getDate(end);
  return `${monthName} ${getDate(start)}${
    startDate !== endDate ? ` - ${endDate}` : ''
  }`;
}

export function getTooltipLabel(
  datum: TrendChartDatum,
  formatValue: ValueFormatter,
  formatOptions?: FormatOptions
) {
  if (!datum.date) {
    return '';
  }
  const date = format(datum.date, 'MMM D YYYY');
  return `${date}: ${formatValue(datum.y, datum.units, formatOptions)}`;
}
