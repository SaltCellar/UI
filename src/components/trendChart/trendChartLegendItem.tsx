import { css } from '@patternfly/react-styles';
import React from 'react';
import { styles } from './trendChartLegendItem.styles';
import { getDateRangeString, TrendChartDatum } from './trendChartUtils';

interface TrendChartLegendItemProps {
  data: TrendChartDatum[];
  isCurrent?: boolean;
}

const TrendChartLegendItem: React.SFC<TrendChartLegendItemProps> = ({
  data,
  isCurrent,
}) => {
  if (!data || data.length === 0) {
    return <div />;
  }
  const range = getDateRangeString(data);
  return (
    <div className={css(styles.trendlegendItem)}>
      <div
        className={css(
          styles.color,
          isCurrent ? styles.currentColor : styles.previousColor
        )}
      />
      {range}
    </div>
  );
};

export { TrendChartLegendItem, TrendChartLegendItemProps };
