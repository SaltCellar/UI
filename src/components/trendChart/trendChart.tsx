import { css } from '@patternfly/react-styles';
import { Report } from 'api/reports';
import React from 'react';
import { FormatOptions, ValueFormatter } from 'utils/formatValue';
import {
  VictoryArea,
  VictoryGroup,
  VictoryTooltip,
  VictoryVoronoiContainer,
} from 'victory';
import { chartStyles, styles } from './trendChart.styles';
import { TrendChartLegend } from './trendChartLegend';
import { TrendChartLegendItem } from './trendChartLegendItem';
import { TrendChartTitle } from './trendChartTitle';
import {
  getTooltipLabel,
  transformReport,
  TrendChartDatum,
  TrendChartType,
} from './trendChartUtils';

interface TrendChartProps {
  title: string;
  height: number;
  current: Report;
  previous: Report;
  type: TrendChartType;
  formatDatumValue: ValueFormatter;
  formatDatumOptions?: FormatOptions;
}

interface State {
  width: number;
}

class TrendChart extends React.Component<TrendChartProps, State> {
  private containerRef = React.createRef<HTMLDivElement>();
  public state: State = {
    width: 0,
  };

  public shouldComponentUpdate(nextProps: TrendChartProps) {
    if (!nextProps.current || !nextProps.previous) {
      return false;
    }
    return true;
  }

  private getTooltipLabel = (datum: TrendChartDatum) => {
    const { formatDatumValue, formatDatumOptions } = this.props;
    return getTooltipLabel(datum, formatDatumValue, formatDatumOptions);
  };

  private handleResize = () => {
    this.setState({ width: this.containerRef.current.clientWidth });
  };

  public componentDidMount() {
    setTimeout(() => {
      this.setState({ width: this.containerRef.current.clientWidth });
      window.addEventListener('resize', this.handleResize);
    });
  }

  public componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
  }

  public render() {
    const { title, current, previous, height, type } = this.props;

    const currentData = transformReport(current, type);
    const previousData = transformReport(previous, type);

    return (
      <div className={css(styles.reportSummaryTrend)} ref={this.containerRef}>
        <VictoryGroup
          padding={chartStyles.padding}
          style={chartStyles.group}
          height={height}
          width={this.state.width}
          containerComponent={
            <VictoryVoronoiContainer
              voronoiDimension="x"
              title={title}
              responsive={false}
              labels={this.getTooltipLabel}
              labelComponent={
                <VictoryTooltip
                  cornerRadius={0}
                  style={chartStyles.tooltipText}
                  flyoutStyle={chartStyles.tooltipFlyout}
                />
              }
            />
          }
        >
          {Boolean(previousData.length) && (
            <VictoryArea
              style={chartStyles.previousMonth}
              data={previousData}
            />
          )}
          {Boolean(currentData.length) && (
            <VictoryArea style={chartStyles.currentMonth} data={currentData} />
          )}
        </VictoryGroup>
        <TrendChartTitle>{title}</TrendChartTitle>
        <TrendChartLegend>
          <TrendChartLegendItem isCurrent data={currentData} />
          <TrendChartLegendItem data={previousData} />
        </TrendChartLegend>
      </div>
    );
  }
}

export { TrendChart, TrendChartProps };
