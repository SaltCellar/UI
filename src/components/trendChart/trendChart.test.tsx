jest.mock('date-fns/format');

import { Report, ReportData } from 'api/reports';
import { ChartLegendItem, ChartTitle } from 'components/commonChart';
import * as utils from 'components/commonChart/chartUtils';
import formatDate from 'date-fns/format';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import {
  VictoryArea,
  VictoryGroup,
  VictoryVoronoiContainerProps,
} from 'victory';
import { TrendChart, TrendChartProps } from './trendChart';

const currentMonthReport: Report = createReport('1-15-18');
const previousMonthReport: Report = createReport('12-15-17');

jest.spyOn(utils, 'transformReport');
jest.spyOn(utils, 'getTooltipLabel');

const transformReport = utils.transformReport as jest.Mock;
const getTooltipLabel = utils.getTooltipLabel as jest.Mock;

const props: TrendChartProps = {
  title: 'Trend Title',
  height: 100,
  formatDatumValue: jest.fn(),
  current: currentMonthReport,
  previous: previousMonthReport,
  formatDatumOptions: {},
  type: utils.ChartType.rolling,
};

test('reports are formatted to datums', () => {
  const view = shallow(<TrendChart {...props} />);
  expect(transformReport).toBeCalledWith(currentMonthReport, props.type);
  expect(transformReport).toBeCalledWith(previousMonthReport, props.type);
  const charts = view.find(VictoryArea);
  expect(charts.length).toBe(2);
  expect(charts.at(0).prop('data')).toMatchSnapshot('previous month data');
  expect(charts.at(1).prop('data')).toMatchSnapshot('current month data');
});

test('null previous and current reports are handled', () => {
  const view = shallow(
    <TrendChart {...props} current={null} previous={null} />
  );
  expect(transformReport).toBeCalledWith(null, props.type);
  expect(transformReport).toBeCalledWith(null, props.type);
  const charts = view.find(VictoryArea);
  expect(charts.length).toBe(0);
});

test('legends are created for current and previous data', () => {
  const view = shallow(<TrendChart {...props} />);
  const legends = view.find(ChartLegendItem);
  expect(legends.length).toBe(2);
  expect(legends.at(0).prop('data')).toMatchSnapshot('current month data');
  expect(legends.at(1).prop('data')).toMatchSnapshot('previous month data');
});

test('chart title is displayed and passed to chart container', () => {
  const view = shallow(<TrendChart {...props} />);
  expect(view.find(ChartTitle).prop('children')).toBe(props.title);
  expect(getChartContainerProps(view).title).toBe(props.title);
});

test('height from props is used', () => {
  const view = shallow(<TrendChart {...props} />);
  expect(view.find(VictoryGroup).prop('height')).toBe(props.height);
});

test('labels formats with datum and value formatted from props', () => {
  const view = shallow(<TrendChart {...props} />);
  const datum: utils.ChartDatum = {
    x: 1,
    y: 1,
    key: '1-1-1',
    units: 'units',
  };
  getChartContainerProps(view).labels(datum);
  expect(getTooltipLabel).toBeCalledWith(
    datum,
    props.formatDatumValue,
    props.formatDatumOptions,
    'date'
  );
  expect(props.formatDatumValue).toBeCalledWith(
    datum.y,
    datum.units,
    props.formatDatumOptions
  );
  expect(formatDate).toBeCalledWith(datum.key, expect.any(String));
  expect(view.find(VictoryGroup).prop('height')).toBe(props.height);
});

test('labels ignores datums without a date', () => {
  const view = shallow(<TrendChart {...props} />);
  const datum: utils.ChartDatum = {
    x: 1,
    y: 1,
    key: '',
    units: 'units',
  };
  const value = getChartContainerProps(view).labels(datum);
  expect(value).toBe('');
  expect(props.formatDatumValue).not.toBeCalled();
  expect(formatDate).not.toBeCalled();
});

test('trend is a running total', () => {
  const multiDayReport: Report = {
    data: [
      createReportDataPoint('1-15-18', 1),
      createReportDataPoint('1-16-18', 2),
    ],
  };
  const view = shallow(<TrendChart {...props} current={multiDayReport} />);
  const charts = view.find(VictoryArea);
  expect(charts.at(1).prop('data')).toMatchSnapshot('current month data');
});

test('trend is a daily value', () => {
  const multiDayReport: Report = {
    data: [
      createReportDataPoint('1-15-18', 1),
      createReportDataPoint('1-16-18', 2),
    ],
  };
  const view = shallow(
    <TrendChart
      {...props}
      current={multiDayReport}
      type={utils.ChartType.daily}
    />
  );
  const charts = view.find(VictoryArea);
  expect(charts.at(1).prop('data')).toMatchSnapshot('current month data');
});

function getChartContainerProps(
  view: ShallowWrapper
): VictoryVoronoiContainerProps {
  const group = view.find(VictoryGroup);
  return group.props().containerComponent.props;
}

function createReport(date: string): Report {
  return {
    data: [createReportDataPoint(date)],
  };
}

function createReportDataPoint(date: string, total = 1): ReportData {
  return {
    date,
    values: [{ date, total, units: 'unit' }],
  };
}
