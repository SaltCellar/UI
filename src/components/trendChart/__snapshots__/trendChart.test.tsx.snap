// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`legends are created for current and previous data: current month data 1`] = `
Array [
  Object {
    "date": "1-15-18",
    "units": "unit",
    "x": 15,
    "y": 1,
  },
]
`;

exports[`legends are created for current and previous data: previous month data 1`] = `
Array [
  Object {
    "date": "12-15-17",
    "units": "unit",
    "x": 15,
    "y": 1,
  },
]
`;

exports[`reports are formatted to datums: current month data 1`] = `
Array [
  Object {
    "date": "1-15-18",
    "units": "unit",
    "x": 15,
    "y": 1,
  },
]
`;

exports[`reports are formatted to datums: previous month data 1`] = `
Array [
  Object {
    "date": "12-15-17",
    "units": "unit",
    "x": 15,
    "y": 1,
  },
]
`;

exports[`trend is a daily value: current month data 1`] = `
Array [
  Object {
    "date": "1-15-18",
    "units": "unit",
    "x": 15,
    "y": 1,
  },
  Object {
    "date": "1-16-18",
    "units": "unit",
    "x": 16,
    "y": 2,
  },
]
`;

exports[`trend is a running total: current month data 1`] = `
Array [
  Object {
    "date": "1-15-18",
    "units": "unit",
    "x": 15,
    "y": 1,
  },
  Object {
    "date": "1-16-18",
    "units": "unit",
    "x": 16,
    "y": 3,
  },
]
`;
